### DISCUSSION TITLE
Handling environment variables in XRC files

### DISCUSSION THREAD

**Discussion by ClaudeTO80** [wxFormBuilder/wxFormBuilder#847] [Q&A]
Hi all.
I am using wxFormBuilder for a personal test project. 

I have seen that XRC files allow using environment variables. I have succesfully defined by hand a bitmap image (editing an XRC file created by wxFormbBuilder) whose path is parametric with an environment variable. But in wxFormBuilder I didn't see anything allowing its definition (images bitmap files use relative paths). Am I missing something or it's a missing feature in tool?

Thanks for your support!

**Comment by sodevel** [comment]
Until today i didn't know this is possible, i checked the wxWidgets documentation, but information is scarce, how to you actually reference an environment variable, what syntax has to be used? Anyway, i'm not sure what your request for wxFB is actually, environment variables are defined by, well, the environment, so outside of wxFB. Can't you enter environment variables in the required syntax into the path field? Or does the preview not evaluate them so the image isn't displayed there?

**Comment by ClaudeTO80** [comment]
First of all thank you for your prompt reply. 
What happens is that if I set manually path of image containing environment variable, only file name is taken and stored (how shown first attached image)
![image](https://github.com/user-attachments/assets/7322098c-1012-4954-abc2-be3b609109d8)

Next, when trying to show preview, it fails because images are not found. It is absoluty right that wxFB does not have any idea of images' location. I just wonder if it is possibile skip or ignore these messages and show the preview without images. 
![image](https://github.com/user-attachments/assets/34c51aa2-3054-49dd-8942-ce28f64fc681)

Maybe that what I am using is not the best approach and maybe I will improve it. It should be nice if wxFB would be more robust in order to show previews and handle informations that can not be show in preview.

It is not a blocking issue. If nobody required it until now maybe that this fesature in not used or known.
Thanks for all work you all are doing.

**Comment by sodevel** [comment]
I don't know how useful it is for an application to depend on environment variables to load its resources, right now i don't know in what situation this can be useful instead of harmful, but this is another topic. In regard to wxFB this approach is not so useful because this specific feature does ONLY work for XRC and nothing else and XRC is inferior to generated code. But anyway, i gave it a try and discovered two issues. 

First, you cannot enter environment variables into the `file_path` property because its input validation does remove such values. You can circumvent this by using a bitmap type that uses a simple string property for its value like `Load from XRC`. Using such a type preserves the environment variables in the value, while this will lead to wrong generated code this doesn't matter because generated code doesn't work with such paths anyway, such values will be passed as-is into the XRC and this is what is needed.

The second issue is, the XRC preview does not have the usage of environment variables enabled. This results in that error dialog but the preview is loaded anyway, just without images. I made a quick test and enabled the usage of environment variables, you can download the builds from the GitHub Actions (Branch `fix/xrc-envvars`) to see for yourself. This test is not production ready, it fails to compile on older Ubuntu versions because their wxWidgets version is too old, but its a proof of concept.

In general the whole XRC preview function in wxFB is quite bad and should be reworked, but i don't have an ETA for that though.
